Gems/Atom/RHI/Code/Source/RHI/MultiDevicePipelineLibrary.cpp

```diff
@@ -0,0 +1,123 @@
+ *
+ */
+
+#include <Atom/RHI/Factory.h>
+#include <Atom/RHI/MultiDevicePipelineLibrary.h>
+#include <Atom/RHI/PipelineLibrary.h>
+#include <Atom/RHI/RHISystemInterface.h>
+#include <AzFramework/StringFunc/StringFunc.h>
+
+namespace AZ::RHI
+{
+    bool MultiDevicePipelineLibrary::ValidateIsInitialized() const
+    {
+        if (Validation::IsEnabled())
+        {
+            if (!IsInitialized())
+            {
+                AZ_Error(
+                    "MultiDevicePipelineLibrary",
+                    false,
+                    "MultiDevicePipelineLibrary is not initialized. This operation is only permitted on an initialized library.");
+                return false;
+            }
+        }
+        return true;
+    }
+
+    ResultCode MultiDevicePipelineLibrary::Init(MultiDevice::DeviceMask deviceMask, const MultiDevicePipelineLibraryDescriptor& descriptor)
+    {
+        if (Validation::IsEnabled())
+        {
+            if (IsInitialized())
+            {
+                AZ_Error(
+                    "MultiDevicePipelineLibrary",
+                    false,
+                    "MultiDevicePipelineLibrary is initialized. This operation is only permitted on an uninitialized library.");
+                return ResultCode::InvalidOperation;
+            }
+        }
+
+        MultiDeviceObject::Init(deviceMask);
+
+        ResultCode resultCode = ResultCode::Success;
+
+        IterateDevices(
+            [this, &descriptor, &resultCode](int deviceIndex)
+            {
+                auto* device = RHISystemInterface::Get()->GetDevice(deviceIndex);
+
+                m_devicePipelineLibraries[deviceIndex] = Factory::Get().CreatePipelineLibrary();
+                resultCode =
+                    m_devicePipelineLibraries[deviceIndex]->Init(*device, descriptor.GetDevicePipelineLibraryDescriptor(deviceIndex));
+
+                return resultCode == ResultCode::Success;
+            });
+
+        if(resultCode != ResultCode::Success)
+        {
+            // Reset already initialized device-specific PipelineLibraries and set deviceMask to 0
+            m_devicePipelineLibraries.clear();
+            MultiDeviceObject::Init(static_cast<MultiDevice::DeviceMask>(0u));
+        }
+
+
+        return resultCode;
+    }
+
+    ResultCode MultiDevicePipelineLibrary::MergeInto(AZStd::span<const MultiDevicePipelineLibrary* const> librariesToMerge)
+    {
+        if (!ValidateIsInitialized())
+        {
+            return ResultCode::InvalidOperation;
+        }
+
+        ResultCode result = ResultCode::Success;
+
+        for (auto& [deviceIndex, devicePipelineLibrary] : m_devicePipelineLibraries)
+        {
+            AZStd::vector<const PipelineLibrary*> deviceLibrariesToMerge(librariesToMerge.size());
+
+            for (int i = 0; i < librariesToMerge.size(); ++i)
+            {
+                deviceLibrariesToMerge[i] = librariesToMerge[i]->m_devicePipelineLibraries.at(deviceIndex).get();
```

Comment:
So far, we assumed that the user is responsible that the masks fit, but you're right, we should at least validate it. I've now changed it however to be graceful with the masks and handle only those devices, which both use. That's probably a more reasonable approach here, since it's quite possible that specific pipelines will only be available for specific devices.